### ISSUE TITLE
User-defined flags for compile-time behaviour

### ISSUE THREAD

**Issue by mrmr1993** [ocaml/dune#4483]
Over in the [mina project](https://github.com/MinaProtocol/mina) we've been using the dune profile to choose the compile-time configuration to build with, with the help of a [fairly unpleasant hack](https://github.com/MinaProtocol/mina/blob/52965c37c1eb37cf071f548e45e1bfd49dca717f/src/config/dune#L21). This works well, in the sense that we can build with the chosen configuration, but also has a few limitations:
* changing the profile invalidates all build assets, even those that do not depend on the compile-time configuration, so switching between configurations results in a complete recompile
* there is no fine-grained control over the options in the configuration.

To address these limitations, it would be helpful to add the ability to pass custom flags to dune, for example `--custom-flag foo=bar`. The most ideal implementation then might have
* `%{custom_flag:foo}` variables
  - this would be the most powerful if it can also be used in `(library (enabled_if ...))` stanzas, but appearing in normal variable positions would already be a big improvement
* a ppx cookie exposing the custom flags
  - this would be handy for a `ppx_optcomp`-style preprocessor
  - this would presumably also need some way to declare these flags as preprocessor dependencies, and perhaps also to limit the available flags to those listed as dependencies.

I imagine that the hardest part of this will be invalidating the correct build artifacts when changing flags, but hopefully something similar to what's already done with environment variables will work.

I'm happy to work on a prototype for this, if it seems like a reasonable feature to add.

**Comment by jberdine** [comment]
This use case is similar to one I have, where multiple dune contexts are set up, each with a dune profile, and the profiles are configured with different flags and env vars, which determine cookies. See e.g. [dune-workspace](https://github.com/facebook/infer/blob/master/sledge/dune-workspace) for the context declarations and associating them to profiles, the main [dune](https://github.com/facebook/infer/blob/master/sledge/dune) to specify the flags and env vars for each profile, and e.g. a ppx's [dune](https://github.com/facebook/infer/blob/master/sledge/ppx_trace/dune) that sets a cookie based on an env var set in the profile.

The benefit of this approach is that due to having multiple contexts, up-to-date build artifacts can be maintained for all configurations, and they can be built in parallel. On the other hand, it is too complicated, is spread over many config files,  and involves checking in a dune-workspace file with a hard-coded opam switch name. It would be great if there was a simpler way.

**Comment by bobot** [comment]
Dune-workspace are indeed very nice but also too precise (depends on the global user opam) to be committed into a repository. Something seems indeed missing, before 3.0 is the right time to discuss that.

 * What do you mean by "changing the profile invalidates all build assets", do we agree that they are not invalidated in the sense that if you come back to the previous profile they will be used? It is just that the build assets are not shared between two profiles?
 
 * Does the cache is able to reuse the build assets correctly?

 * I don't yet get how you would use the feature. Do you think to some time calling `dune build --custom-flag foo` and some other time `dune build --custom-flag bar`?

 * Using `--custom-flag` without using different workspace will disallow parallel build, is it a problem for you?

 * If you are not interested in parallel build, what is the drawbacks of generating before calling dune a dune file with a `(env ( _ (env-vars "MYVAR" ...)))` stanza?

**Comment by jberdine** [comment]
Maybe related, @jeremiedimino previously [suggested](https://github.com/ocaml/dune/issues/3279#issuecomment-605935829) that perhaps contexts and profiles ought to be combined into one concept for 3.0.

**Comment by mrmr1993** [comment]
> * What do you mean by "changing the profile invalidates all build assets", do we agree that they are not invalidated in the sense that if you come back to the previous profile they will be used? It is just that the build assets are not shared between two profiles?
> * Does the cache is able to reuse the build assets correctly?

When I switch build profiles, the build artifacts are recompiled, with the old ones being overwritten. This means that it's also not possible for dune to switch back to them, because they've been replaced. Conflating build profiles with contexts is also not ideal for me: creating multiple copies of (IIRC, ~20GB of) mostly identical build assets is just another thing I would need to hack around.

For the cache, I'm currently unable to use it because we've committed the cardinal sin of non-deterministic build assets.

My issue effectively boils down to: when I switch build profiles, I spend a bunch of time recompiling things that I already have perfectly good versions of, but I have no good options for configuration other than build profiles.

> * I don't yet get how you would use the feature. Do you think to some time calling `dune build --custom-flag foo` and some other time `dune build --custom-flag bar`?
> * Using `--custom-flag` without using different workspace will disallow parallel build, is it a problem for you?

Exactly that, yes. I don't think I'm using parallel builds, and it seems to me that environment variables -- which we also use -- must have the same issue.

> * If you are not interested in parallel build, what is the drawbacks of generating before calling dune a dune file with a `(env ( _ (env-vars "MYVAR" ...)))` stanza?

It seems like this requires a different build system to wrap dune to make it do the right thing. I also find that environment variables are error-prone, in that you may `export FOO=bar` and later forget, misconfiguring all subsequent builds. I would prefer to use them sparingly if possible.

If this configurability isn't a goal of dune, that's fine and I can wrap it with a build system that does support it, but it's a nuisance that dune can't handle this itself.

**Comment by bobot** [comment]
I indeed mixed-up `build contexts` and `profiles`, I though you were using the former. It is possible that profiles are used directly as memoization key, I agree it seems like a misfeature.

Your proposition seems reasonable, I'm just trying to understand your need better. For example I don't understand the fairly unpleasant hack you pointed. Why do you have this dependency on all the files in the directory that doesn't have the extension `un~`, when the rule is just a copy. Is it linked to the use of `%{profile}` ?

Design and implementation choices to make:
  - could `custom-flag` be set in the dune-workspace?
  - why will it not have the problem of `profile` ?
  - where could the variable appear?
  - `dune build --set-env FOO=bar` could also be used for setting environment variable

I will let other developpers chime in.

**Comment by aalekseyev** [comment]
I read the whole discussion, but I still don't understand the problems and how the proposed change is supposed to solve them.

@mrmr1993, you listed two bullet-points as problems (excessive rebuilds and insufficient fine-grained control).
Can you expand a bit more on each of these, with examples?
What artifacts concretely get rebuild and how/why you think their rebuild can in principle be avoided?

What fine-grained control is needed and why can't it be gotten through various dune features? What have you tried?

A couple of general thoughts:

- Dune is primarily configured through files. You're giving the source files as files and the dune files as files, so I don't see why it's such a tragedy to have to configure another thing via a file as in @bobot's proposal.

- Declaring and re-using variables in dune can be awkward and this problem keeps coming up, so it's definitely on our radar. We can't jump to any specific solution too hastily, though, as it's easy to design these things badly and get stuck with them (basically like programming language features), so we're being careful about it.
